File: data/repositories/flatpak_repository.py
import gi
gi.require_version('Flatpak', '1.0')
from gi.repository import Flatpak
import requests
from typing import List
from domain.entities import Package
from domain.repositories import PackageRepository

class FlatpakPackageRepository(PackageRepository):
    def __init__(self):
        pass

    def _get_installed_flatpaks(self) -> dict:
        installed_flatpaks = {}
        for inst_func in [Flatpak.Installation.new_system, Flatpak.Installation.new_user]:
            try:
                inst = inst_func(None)
                for ref in inst.list_installed_refs(None):
                    if ref.get_kind() == Flatpak.RefKind.APP:
                        installed_flatpaks[ref.get_name()] = ref.get_appdata_version() or ""
            except Exception:
                pass
        return installed_flatpaks

    def search(self, query: str) -> List[Package]:
        if not query:
            return []
            
        results = []
        try:
            r = requests.post("https://flathub.org/api/v2/search", json={"query": query})
            if r.status_code == 200:
                hits = r.json().get("hits", [])
                installed_flatpaks = self._get_installed_flatpaks()
                
                for hit in hits[:30]:
                    app_id = hit.get("app_id") or hit.get("id")
                    if not app_id:
                        continue
                        
                    installed = app_id in installed_flatpaks
                    results.append(Package(
                        name=app_id,
                        title=hit.get("name") or app_id,
                        desc=hit.get("summary") or "",
                        version=hit.get("version") or "",
                        type="flatpak",
                        installed=installed,
                        installed_version=installed_flatpaks.get(app_id, "") if installed else "",
                        icon=hit.get("icon") or ""
                    ))
        except Exception as e:
            print(f"Error searching Flatpak: {e}")
            
        return results

    def get_installed(self) -> List[Package]:
        installed = []
        for inst_func in [Flatpak.Installation.new_system, Flatpak.Installation.new_user]:
            try:
                inst = inst_func(None)
                for ref in inst.list_installed_refs(None):
                    if ref.get_kind() == Flatpak.RefKind.APP:
                        installed.append(Package(
                            name=ref.get_name(),
                            title=ref.get_appdata_name() or ref.get_name(),
                            desc=ref.get_appdata_summary() or "",
                            version=ref.get_appdata_version() or "",
                            type="flatpak",
                            installed=True,
                            installed_version=ref.get_appdata_version() or ""
                        ))
            except Exception as e:
                print(f"Error loading installed flatpak: {e}")
        return installed

    def is_installed(self, app_id: str) -> bool:
        return app_id in self._get_installed_flatpaks()

    def get_popular(self) -> List[Package]:
        results = []
        try:
            r = requests.get("https://flathub.org/api/v2/collection/popular", timeout=5)
            if r.status_code == 200:
                hits = r.json().get("hits", [])
                installed_flatpaks = self._get_installed_flatpaks()
                
                for hit in hits[:12]:
                    app_id = hit.get("app_id") or hit.get("id")
                    if not app_id:
                        continue
                    installed = app_id in installed_flatpaks
                    results.append(Package(
                        name=app_id,
                        title=hit.get("name") or app_id,
                        desc=hit.get("summary") or "",
                        version=hit.get("version") or "latest",
                        type="flatpak",
                        installed=installed,
                        installed_version=installed_flatpaks.get(app_id, "") if installed else "",
                        icon=hit.get("icon") or ""
                    ))
        except Exception as e:
            print(f"Error fetching popular flatpaks: {e}")
            
        return results

    def get_install_command(self, pkg_name: str) -> List[str]:
        # Robust command: ensure flathub remote is added and install from it explicitly
        return [
            "sh", "-c", 
            f"flatpak remote-add --user --if-not-exists flathub https://flathub.org/repo/flathub.flatpakrepo && "
            f"flatpak install --user -y flathub '{pkg_name}'"
        ]

    def get_uninstall_command(self, pkg_name: str) -> List[str]:
        # Try to uninstall from user first, then system (if possible without sudo, though system usually needs it)
        # But for consistency with install, we prioritize --user
        return ["flatpak", "uninstall", "--user", "-y", pkg_name]

    def get_updatable(self) -> List[Package]:
        updates = []
        for inst_func in [Flatpak.Installation.new_system, Flatpak.Installation.new_user]:
            try:
                inst = inst_func(None)
                for ref in inst.list_installed_refs(None):
                    if ref.get_kind() == Flatpak.RefKind.APP:
                        local_commit = ref.get_commit()
                        try:
                            remote_ref = inst.fetch_remote_ref_sync(ref.get_remote(), ref.get_kind(), ref.get_name(), ref.get_arch(), ref.get_branch(), None)
                            if remote_ref:
                                remote_commit = remote_ref.get_commit()
                                if local_commit != remote_commit:
                                    updates.append(Package(
                                        name=ref.get_name(),
                                        title=ref.get_appdata_name() or ref.get_name(),
                                        desc=ref.get_appdata_summary() or "",
                                        version=remote_ref.get_appdata_version() or "latest",
                                        type="flatpak",
                                        installed=True,
                                        installed_version=ref.get_appdata_version() or "",
                                        icon=""
                                    ))
                        except Exception:
                            pass
            except Exception as e:
                print(f"Error checking flatpak updates: {e}")
        return updates

File: src-rust/src/data/flatpak_repository.rs
use std::collections::HashMap;
use async_trait::async_trait;
use serde::Deserialize;
use crate::domain::entities::Package;
use crate::domain::repositories::PackageRepository;

pub struct FlatpakPackageRepository {
    client: reqwest::Client,
}

impl FlatpakPackageRepository {
    pub fn new() -> Self {
        Self {
            client: reqwest::Client::builder()
                .timeout(std::time::Duration::from_secs(10))
                .build()
                .expect("Failed to build reqwest client"),
        }
    }

    /// Returns a map of { app_id -> version } for all installed flatpaks.
    pub async fn get_installed_flatpaks(&self) -> HashMap<String, String> {
        tokio::task::spawn_blocking(|| {
            let output = std::process::Command::new("flatpak")
                .args(["list", "--app", "--columns=application,version"])
                .output();

            let mut map = HashMap::new();
            if let Ok(out) = output {
                let text = String::from_utf8_lossy(&out.stdout);
                for line in text.lines() {
                    let parts: Vec<&str> = line.splitn(2, '\t').collect();
                    if parts.len() == 2 {
                        map.insert(parts[0].trim().to_string(), parts[1].trim().to_string());
                    }
                }
            }
            map
        })
        .await
        .unwrap_or_default()
    }

    #[allow(dead_code)]
    pub async fn is_installed(&self, app_id: &str) -> bool {
        self.get_installed_flatpaks().await.contains_key(app_id)
    }

    pub async fn get_popular(&self) -> Vec<Package> {
        let installed = self.get_installed_flatpaks().await;
        let url = "https://flathub.org/api/v2/collection/popular";
        match self.client.get(url).send().await {
            Ok(resp) => {
                if let Ok(data) = resp.json::<FlathubCollectionResponse>().await {
                    data.hits.into_iter().take(12).filter_map(|hit| {
                        let app_id = hit.app_id.or(hit.id)?;
                        let is_installed = installed.contains_key(&app_id);
                        Some(Package {
                            installed_version: if is_installed {
                                installed.get(&app_id).cloned().unwrap_or_default()
                            } else {
                                String::new()
                            },
                            name: app_id.clone(),
                            title: hit.name.unwrap_or_else(|| app_id.clone()),
                            desc: hit.summary.unwrap_or_default(),
                            version: hit.version.unwrap_or_else(|| "latest".into()),
                            pkg_type: "flatpak".to_string(),
                            installed: is_installed,
                            icon: hit.icon.unwrap_or_default(),
                        })
                    }).collect()
                } else {
                    vec![]
                }
            }
            Err(e) => {
                eprintln!("Error fetching popular flatpaks: {}", e);
                vec![]
            }
        }
    }

    pub async fn get_updatable(&self) -> Vec<Package> {
        // flatpak remote-ls with update check is slow; use `flatpak update --dry-run`
        tokio::task::spawn_blocking(|| {
            let output = std::process::Command::new("flatpak")
                .args(["update", "--dry-run", "--app", "--columns=application,version"])
                .output();

            let mut updates = vec![];
            if let Ok(out) = output {
                let text = String::from_utf8_lossy(&out.stdout);
                for line in text.lines() {
                    let parts: Vec<&str> = line.splitn(2, '\t').collect();
                    if parts.len() == 2 {
                        let app_id = parts[0].trim().to_string();
                        let new_version = parts[1].trim().to_string();
                        updates.push(Package {
                            name: app_id.clone(),
                            title: app_id.clone(),
                            desc: String::new(),
                            version: new_version,
                            pkg_type: "flatpak".to_string(),
                            installed: true,
                            installed_version: String::new(),
                            icon: String::new(),
                        });
                    }
                }
            }
            updates
        })
        .await
        .unwrap_or_default()
    }
}

#[async_trait]
impl PackageRepository for FlatpakPackageRepository {
    async fn search(&self, query: &str) -> Vec<Package> {
        let installed = self.get_installed_flatpaks().await;
        let payload = serde_json::json!({"query": query});

        match self.client
            .post("https://flathub.org/api/v2/search")
            .json(&payload)
            .send()
            .await
        {
            Ok(resp) => {
                if let Ok(data) = resp.json::<FlathubSearchResponse>().await {
                    data.hits.into_iter().take(30).filter_map(|hit| {
                        let app_id = hit.app_id.or(hit.id)?;
                        let is_installed = installed.contains_key(&app_id);
                        Some(Package {
                            installed_version: if is_installed {
                                installed.get(&app_id).cloned().unwrap_or_default()
                            } else {
                                String::new()
                            },
                            name: app_id.clone(),
                            title: hit.name.unwrap_or_else(|| app_id.clone()),
                            desc: hit.summary.unwrap_or_default(),
                            version: hit.version.unwrap_or_default(),
                            pkg_type: "flatpak".to_string(),
                            installed: is_installed,
                            icon: hit.icon.unwrap_or_default(),
                        })
                    }).collect()
                } else {
                    vec![]
                }
            }
            Err(e) => {
                eprintln!("Flatpak search error: {}", e);
                vec![]
            }
        }
    }

    async fn get_installed(&self) -> Vec<Package> {
        tokio::task::spawn_blocking(|| {
            let output = std::process::Command::new("flatpak")
                .args([
                    "list", "--app",
                    "--columns=application,name,description,version",
                ])
                .output();

            let mut pkgs = vec![];
            if let Ok(out) = output {
                let text = String::from_utf8_lossy(&out.stdout);
                for line in text.lines() {
                    let parts: Vec<&str> = line.splitn(4, '\t').collect();
                    if parts.is_empty() { continue; }
                    let app_id = parts[0].trim().to_string();
                    let title   = parts.get(1).map(|s| s.trim().to_string()).unwrap_or_else(|| app_id.clone());
                    let desc    = parts.get(2).map(|s| s.trim().to_string()).unwrap_or_default();
                    let version = parts.get(3).map(|s| s.trim().to_string()).unwrap_or_default();
                    pkgs.push(Package {
                        name: app_id.clone(),
                        title,
                        desc,
                        version: version.clone(),
                        pkg_type: "flatpak".to_string(),
                        installed: true,
                        installed_version: version,
                        icon: String::new(),
                    });
                }
            }
            pkgs
        })
        .await
        .unwrap_or_default()
    }

    async fn get_install_command(&self, pkg_name: &str) -> Vec<String> {
        vec![
            "sh".into(), "-c".into(),
            format!(
                "flatpak remote-add --user --if-not-exists flathub \
                 https://flathub.org/repo/flathub.flatpakrepo && \
                 flatpak install --user -y flathub '{}'",
                pkg_name
            ),
        ]
    }

    async fn get_uninstall_command(&self, pkg_name: &str) -> Vec<String> {
        vec!["flatpak".into(), "uninstall".into(), "--user".into(), "-y".into(), pkg_name.to_string()]
    }
}

// ─── Flathub API types ───────────────────────────────────────────────────────

#[derive(Deserialize)]
struct FlathubSearchResponse {
    hits: Vec<FlathubHit>,
}

#[derive(Deserialize)]
struct FlathubCollectionResponse {
    hits: Vec<FlathubHit>,
}

#[derive(Deserialize)]
struct FlathubHit {
    app_id: Option<String>,
    id: Option<String>,
    name: Option<String>,
    summary: Option<String>,
    version: Option<String>,
    icon: Option<String>,
}
